File: src/dbscan.py
import numpy as np
import logging
import time 
import os
import pandas as pd
from src.datasets import Dataset
import json
import math

def epsilon_prim(epsilon):
    return math.sqrt(2-2*epsilon)

def setup_logger(name, log_file, level=logging.INFO):
    """To setup as many loggers as you want"""
    formatter = logging.Formatter(fmt='%(msecs)06f,%(message)s')
    handler = logging.FileHandler(log_file)        
    handler.setFormatter(formatter)

    logger = logging.getLogger(name)
    logger.setLevel(level)
    logger.addHandler(handler)

    return logger

logger = setup_logger('root', 'out.log')

def get_neighbors(X, current_index, epsilon, similarity):
    '''
    X - dataset with point coordinates,
    current_index - index of given point,
    epsilon - max similarity or distance to given point,
    similarity - metric of similarity or distance bewteen two points
    '''
    neighbor_indices = []
    for neighbor_index, neighbor in enumerate(X):
        if similarity(neighbor, X[current_index]) <= epsilon:
            neighbor_indices.append(neighbor_index)
    return neighbor_indices

def dbscan(X, epsilon, minPts, similarity):
    '''
    X - dataset with point coordinates,
    epsilon - max similarity or distance to given point,
    minPts - minimum number of points that create cluster,
    similarity - metric of similarity or distance bewteen two points
    '''
    
    # each data point can be in one of 3 stages
    NOT_VISITED = -1 # not visited point
    VISITED = 0 # non-core point
    CLUSTERED = 1 # core point
    
    # initial setup
    n = X.shape[0]
    cluster = np.array([-1] * n) # cluster register
    state = np.array([NOT_VISITED] * n) # state register
    cluster_id = 1

    def search(current_index, cluster_id, epsilon, minPts, similarity):
        ''' Extend cluster 
        current_index - the point for which the search of Epsilon- neighborhood is being made,
        cluster_id - number of the cluster that is currently being extended
        '''
        # calculation of Eps-neighborhood for current_index
        timer1 = time.time() 
        neighbor_indices = get_neighbors(X, current_index, epsilon, similarity)
        logger.info(f'Eps_time,{current_index},{(time.time() - timer1)*1000},')
        logger.info(f'|Eps_neighbors|,{current_index}, {len(neighbor_indices)},')
        neighbor_indices_str =';'.join(str(e) for e in neighbor_indices)
        logger.info(f'Eps_neighbor_id,{current_index},,{neighbor_indices_str}')
        
        if len(neighbor_indices) >= minPts:
            state[current_index] = CLUSTERED
            cluster[current_index] = cluster_id
            for neighbor_index in neighbor_indices:
                if (state[neighbor_index] == VISITED) or (state[neighbor_index] == NOT_VISITED):
                    state[neighbor_index] = CLUSTERED
                    cluster[neighbor_index] = cluster_id
                    
                    # number of distance/similarity calculations
                    logger.info(f'similarity_calculation, {current_index}, 1,')
                    search(neighbor_index, cluster_id, epsilon, minPts, similarity)
        else:
            state[current_index] = VISITED

    while NOT_VISITED in state:
        not_visited_ids = np.where(state==NOT_VISITED)[0]
        search(not_visited_ids[0], cluster_id, epsilon, minPts, similarity)
        cluster_id += 1
    logger.info(f'stop log,,,')
    return cluster, state

def setup_logger(name, log_file, level=logging.INFO):
    """To setup as many loggers as you want"""
    formatter = logging.Formatter(fmt='%(msecs)06f,%(message)s')
    handler = logging.FileHandler(log_file)        
    handler.setFormatter(formatter)

    logger = logging.getLogger(name)
    logger.setLevel(level)
    logger.addHandler(handler)

    return logger

class DBSCAN:

    def __init__(self, epsilon, minPts, similarity, **kwargs):
        self.epsilon = epsilon_prim(epsilon)
        self.minPts = minPts
        self.distance = similarity
        self.log_output = 'out.log'
        self.name = 'dbscan'
        
        config_path = './configs/dbscan.json'
        f = open(config_path)
        self.config = json.load(f)
    
    def run(self, dataset_name):
            
        logger = logging.getLogger('root')
        formatter = logging.Formatter(fmt='%(msecs)06f,%(message)s')    
        handler = logging.FileHandler(self.log_output)
        handler.setFormatter(formatter)
        logger.addHandler(handler)         
        
        logger.info(f'start log,,,')
        
        timer1 = time.time() 
        dataset = Dataset(self.config[dataset_name]['path'])
        X, y = dataset.X, dataset.y
        logger.info(f'reading_data,,{(time.time() - timer1)*1000},')
        
        self.X = X
        result = dbscan(self.X, self.epsilon, self.minPts, self.distance)
        self.y_pred, self.state = result
        
        timer1 = time.time() 
        pd.DataFrame({
            'cluster': self.y_pred,
            'state': self.state
        }).to_csv(f"out/{self.name}_{dataset_name}.csv", header=None)
        logger.info(f'writing_data,,{(time.time() - timer1)*1000},')
        
        return self.y_pred
    
    def get_logs(self):
        logs = pd.read_csv(
            self.log_output,
            names=['time [ms]', 'operation', 'point_id', 'value', 'string']
        )
        # logs['time [ms]'] -= logs['time [ms]'].min()
        return logs
    

File: src/dbscan.cpp
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xview.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xfixed.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xtensor.hpp>
#include <xtensor/xadapt.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xaxis_slice_iterator.hpp>
#include <xtensor/xaxis_iterator.hpp>
#include <xtensor/xio.hpp>
#include <xtensor/xarray.hpp>
#include <xtensor/xindex_view.hpp>
#include <xtensor/xchunked_array.hpp>
#include <xtensor/xfixed.hpp>
#include <cstddef>
#include <vector>
#include <xtensor/xarray.hpp>
#include <xtensor/xadapt.hpp>
#include <istream>
#include <fstream>
#include <iostream>
#include <xtensor/xarray.hpp>
#include <xtensor/xcsv.hpp>
#include <iostream>
#include <fstream>
#include <functional>
#include <cstdio>
#include <xtensor/xcsv.hpp>
#include <iostream>
#include <chrono>
#include <unistd.h>
#include <fstream>
#include <nlohmann/json.hpp>
#include <chrono>
#include <list>
#include <string>
#include <regex>

    
using namespace xt;
using namespace std;
using namespace std::chrono;
using json = nlohmann::json;


double current_time(){
    auto current_time = std::chrono::system_clock::now();
    auto duration_in_seconds = std::chrono::duration<double>(current_time.time_since_epoch());
    double num_seconds = duration_in_seconds.count();
    return num_seconds;
}

float euclidean_distance(const xarray<double>& a, const xarray<double>& b){ 
    return (float) sqrt(sum(pow((a - b), 2))(0));
}

double epsilon_prim(double epsilon){ 
    return (double) sqrt(2 - 2 * epsilon);
}

auto normalize(const xarray<double>& a){ 
    xarray<double> result = a / sqrt(sum(pow(a, 2)));
    return result;
}

xarray<double> CLUSTER; 
xarray<double> STATE;
    

int main() {
    
    string path;
    double epsilon;
    int minPts;
    string out_path;
    string log_out;
    
    // read configs
    std::ifstream file("configs/dbscan.json");
    json config = json::parse(file);
    
    for (auto conf : config)
    {
        if (not conf["disable"].get<bool>()) {
            std::cout << conf["name"].get<std::string>() << "\n";
            auto path = conf["path"].get<std::string>();
            auto epsilon = conf["params_dbscan"]["epsilon"].get<double>();
            auto minPts = conf["params_dbscan"]["minPts"].get<int>();
            auto out_path = conf["out_path"].get<std::string>();
            auto log_out = conf["log_out"].get<std::string>();
            
            epsilon = epsilon_prim(epsilon);
                
            out_path = regex_replace(out_path, regex("algorithm"), "dbscan_cpp");
            log_out = regex_replace(log_out, regex("algorithm"), "dbscan_cpp");
            
            // load data
            ofstream outfile;
            outfile.open(log_out);

            auto start =  chrono::high_resolution_clock::now();

            ifstream input_file;
            input_file.open (path);
            auto data = xt::load_csv<double>(input_file);
            input_file.close();

            auto end =  chrono::high_resolution_clock::now();
            duration<double> elapsed = (end - start);
            outfile << std::setprecision (16) << current_time() << ",reading_data," <<
                "," << elapsed.count() << "," << endl;

            data.reshape(data.shape());
            xarray<double> X = view(data, all(), range(0, -1));

            // init
            void dbscan(const xarray<float>& X, float epsilon, int minPts, std::ofstream& outfile);

            // run algorithm
            std::cout << "Processing ..." << "\n";
            dbscan(X, epsilon, minPts, outfile);
            start = chrono::high_resolution_clock::now();

            // save output
            ofstream out_file;
            out_file.open(out_path);
            dump_csv(out_file, xt::stack(xtuple(CLUSTER, STATE), 1));
            out_file.close();

            elapsed = (chrono::high_resolution_clock::now() - start);
            outfile << std::setprecision (16) << current_time() << ",writing_data," <<
                "," << elapsed.count() << ',' << endl;
            outfile.close();
            std::cout << "Done." << "\n";
        }
    }
}


void dbscan(const xarray<float>& X_original, float epsilon, int minPts, std::ofstream& outfile)
{
    outfile << std::setprecision (16) << current_time() << ",start log,,," << endl;

    // normalization
    auto start =  chrono::high_resolution_clock::now();
    xarray<float> X = normalize(X_original);
    auto end =  chrono::high_resolution_clock::now();
    std::chrono::duration<double> elapsed = end - start;

    outfile << std::setprecision (16) << current_time() << ",normalization_time," <<
        "," << elapsed.count() << "," << endl;
    
    // each data point can be in one of 3 stages
    int NOT_VISITED = -1; // not visited point
    int VISITED = 0; // non-core point
    int CLUSTERED = 1; // core point

    // initial setup
    int n = X.shape()[0];
    xarray<double> cluster = ones<double>({n}) * -1; // cluster register
    xarray<double> state = ones<double>({n}) * NOT_VISITED; // state register
    int cluster_id = 1;
        
    auto get_neighbors = [&] (int current_index) {
        list<int> neighbor_indices = list<int>();
        for (int neighbor_index : arange(0, n)){
            
            outfile << std::setprecision (16) << current_time() << ",similarity_calculation," <<
                current_index << ",1," << endl;
            
            if (euclidean_distance(row(X, neighbor_index), row(X, current_index)) <= (float) epsilon){
                neighbor_indices.push_back((int) neighbor_index);
            };
        }
        return neighbor_indices;
    };
    
    // extend cluster
    function < void( int ) > search = [&] (int current_index) -> void {
        auto start =  chrono::high_resolution_clock::now();
        list<int> neighbor_indices = get_neighbors(current_index);
        auto end = chrono::high_resolution_clock::now();
        std::chrono::duration<double> elapsed = end - start;
        
        outfile << std::setprecision (16) << current_time() << ",Eps_time," <<
                current_index << "," << elapsed.count() << ',' << endl;
        outfile << std::setprecision (16) << current_time() << ",|Eps_neighbors|," <<
                current_index << "," << (int) neighbor_indices.size() << ',' << endl;
        outfile << std::setprecision (16) << current_time() << ",Eps_neighbor_id," <<
                current_index << ",,";
        for(auto& i : neighbor_indices) {outfile << i << ";"; }
        outfile << endl;

        if ((int) neighbor_indices.size() >= minPts){
            state(current_index) = CLUSTERED;
            cluster(current_index) = cluster_id;
            for (auto& neighbor_index : neighbor_indices) {
                if (state(neighbor_index) == VISITED || state(neighbor_index) == NOT_VISITED){
                    state(neighbor_index) = CLUSTERED;
                    cluster(neighbor_index) = cluster_id;
                    search(neighbor_index);
                };
            };
        } else {
            state(current_index) = VISITED;
        };
    };

    while (any(equal(state, NOT_VISITED))){
        xarray<int> idx = from_indices(argwhere(equal(state, NOT_VISITED)));
        search(idx(0,0));
        cluster_id++;
    }

    outfile << std::setprecision (16) << current_time() << ",stop log,,," << endl;
    CLUSTER = cluster;
    STATE = state;
}
